service/r2/service.go

```diff
@@ -553,3 +565,33 @@ type ruleSetJSON struct {
 	MappingRules  []mappingRuleJSON `json:"mappingRules"`
 	RollupRules   []rollupRuleJSON  `json:"rollupRules"`
 }
+
+// Creates a new RuleSetSnapshot from a rulesetJSON. If the ruleSetJSON has no IDs for any of its
+// mapping rules or rollup rules, it generates missing IDs and sets as a string UUID string so they
+// can be stored in a mapping (id -> rule).
+func (r ruleSetJSON) ruleSetSnapshot() *rules.RuleSetSnapshot {
+	rss := rules.RuleSetSnapshot{
+		Namespace:    r.Namespace,
+		Version:      r.Version,
+		MappingRules: map[string]*rules.MappingRuleView{},
```

Comment:
nit: it's more efficient to do the following to avoid repeated map reallocations.
```
mappingRules := make(map[string]*rules.MappingRuleView, len(r.MappingRules))
for _, mr := r.MappingRules {
  ...
}

rollupRules := make(map[string]*rules.RollupRuleView, len(r.RollupRules))
for _, rr := range r.RollupRules {
  ..
}
return rules.RuleSetSnapshot{
  Namespace: r.Namespace,
  Version: r.Version,
  MappingRules: mappingRules,
  RollupRules: rollupRules,
}
```